File: src/simulation/lifecycle/TileMutationHooks.js
/**
 * TileMutationHooks — centralized cleanup that must fire whenever a tile's
 * type changes mid-simulation (sabotage RUINS, demolition, wildfire, build,
 * LLM-placed structures, etc.).
 *
 * Background: tile mutations cascade into many subsystems — building counts,
 * worker reservations, worker target/path, A* path cache, tileState fertility,
 * and ProcessingSystem cooldown timers. Without a single cleanup pass, a
 * Saboteur turning a FARM into RUINS leaves workers frozen with stale paths
 * and stale reservations on the destroyed tile, while ResourceSystem (which
 * rebuilds buildings counts next tick) runs AFTER WorkerAISystem in
 * SYSTEM_ORDER — so within the sabotage tick workers see stale farm counts.
 *
 * Use `mutateTile(state, ix, iz, newTile)` whenever you change a tile during
 * the simulation. It calls setTile() (which already handles tileState +
 * grid.version + tileStateVersion) and then runs the cascade cleanup.
 */
import { TILE } from "../../config/constants.js";
import { BALANCE } from "../../config/balance.js";
import { setTile, rebuildBuildingStats } from "../../world/grid/Grid.js";

/**
 * Tiles that block pathfinding (or that, when arrived at, no longer match a
 * worker's expected target type). Mutating to one of these requires path /
 * target invalidation. Mutating to anything else (e.g. GRASS→FARM, GRASS→ROAD)
 * leaves the path physically walkable — the worker either keeps walking or
 * re-paths lazily on the next tick via grid.version mismatch. Doing aggressive
 * path-nuking here triggered a long-horizon DevIndex regression because
 * workers re-pathed every time a road or building was placed nearby.
 */
const BLOCKING_TILES = new Set([TILE.RUINS, TILE.WALL, TILE.WATER]);

/**
 * Apply cascade cleanup after a tile changes type. Idempotent — safe to call
 * even if oldTile === newTile (will early-return).
 *
 * @param {import("../../app/types.js").GameState} state
 * @param {number} ix
 * @param {number} iz
 * @param {number} oldTile
 * @param {number} newTile
 */
export function onTileMutated(state, ix, iz, oldTile, newTile) {
  if (oldTile === newTile) return;
  if (!state || !state.grid) return;

  // 1. Rebuild building counts synchronously so any system that runs LATER in
  //    the SAME tick (e.g. WorkerAISystem after WorldEventSystem) sees the
  //    new counts and doesn't pick the destroyed tile as a target.
  state.buildings = rebuildBuildingStats(state.grid);

  // 2. Release any worker reservation on this tile.
  const reservation = state._jobReservation;
  if (reservation && typeof reservation.releaseTile === "function") {
    reservation.releaseTile(ix, iz);
  }

  // 3. Invalidate worker target/path. Two distinct cases:
  //    (a) targetTile points at THIS tile and the new type is not a worker
  //        work-site (was FARM → now RUIN/GRASS) — the worker's intent is
  //        invalidated, force retarget next tick.
  //    (b) path passes through THIS tile AND the new tile is impassable
  //        (RUIN/WALL/WATER) — the worker can't physically walk through.
  //
  //    For non-blocking transitions (GRASS→ROAD, GRASS→FARM), we deliberately
  //    skip path invalidation: setTile() already bumped grid.version so the
  //    next hasActivePath() call refreshes lazily, and the path is still
  //    physically walkable in the meantime. Aggressive eager invalidation
  //    here caused workers to re-path every time a tile was built nearby,
  //    measurably degrading long-horizon throughput.
  const newBlocks = BLOCKING_TILES.has(newTile);
  const agents = Array.isArray(state.agents) ? state.agents : [];
  for (const agent of agents) {
    if (!agent || agent.alive === false) continue;
    const target = agent.targetTile;
    const targetMatches = target && target.ix === ix && target.iz === iz;
    let pathBlockedHere = false;
    if (newBlocks && Array.isArray(agent.path)) {
      for (let i = 0; i < agent.path.length; i += 1) {
        const step = agent.path[i];
        if (step && step.ix === ix && step.iz === iz) {
          pathBlockedHere = true;
          break;
        }
      }
    }
    if (targetMatches || pathBlockedHere) {
      if (targetMatches) agent.targetTile = null;
      agent.path = null;
      agent.pathIndex = 0;
      agent.pathGridVersion = -1;
      agent.pathTrafficVersion = 0;
      // v0.8.6 Tier 2 F5: also zero the agent's desired velocity. Without
      // this the agent kept drifting toward the now-invalid path's last
      // computed direction for ≥1 tick (300ms+), occasionally walking onto
      // the just-mutated blocking tile and getting wedged.
      if (agent.desiredVel) {
        agent.desiredVel.x = 0;
        agent.desiredVel.z = 0;
      } else {
        agent.desiredVel = { x: 0, z: 0 };
      }
      if (agent.blackboard) {
        agent.blackboard.pendingPathWorkerKey = "";
        agent.blackboard.pendingPathTargetTile = null;
      }
    }
  }

  // 4. Mark this tile-key as destroyed so ProcessingSystem (and any other
  //    keyed-by-tile cache) can lazily drop its entry on next tick. We use a
  //    Set rather than direct mutation because ProcessingSystem instances are
  //    not stored on state. ProcessingSystem drains this set each tick.
  const tileKey = `${ix},${iz}`;
  if (!state._tileMutationDirtyKeys) {
    state._tileMutationDirtyKeys = new Set();
  }
  state._tileMutationDirtyKeys.add(tileKey);

  // 5. v0.8.4 strategic walls + GATE (Agent C) — wallHp / gateHp lifecycle.
  //    When a tile *becomes* a WALL or GATE, seed the per-tile hp pool so
  //    AnimalAISystem / VisitorAISystem can attack it. When a tile *was* a
  //    WALL or GATE and is no longer, clear the pool so a future build on
  //    the same coords doesn't inherit stale hp. Per-tile state lives on
  //    grid.tileState.get(idx) — the same map setTile() seeds with
  //    fertility / yieldPool / nodeFlags.
  //
  //    setTile preserves a tileState entry for WALL (lumped with ROAD/
  //    BRIDGE) but does NOT have a branch for GATE (a v0.8.4 addition); a
  //    GATE placement currently falls through to the delete branch unless
  //    nodeFlags/yieldPool happen to be non-zero. To keep this hook
  //    self-contained — and to avoid editing Grid.js (broad ownership) —
  //    we lazily create a tileState entry here when wallHp needs to be
  //    seeded. Same `Map<idx, entry>` shape; downstream consumers
  //    (rendering, attack logic, save/load) read wallHp via .get(idx).
  const idx = ix + iz * state.grid.width;
  if (newTile === TILE.WALL || newTile === TILE.GATE) {
    const tileState = state.grid.tileState;
    if (tileState) {
      let entry = tileState.get(idx);
      if (!entry) {
        // Lazily create a minimal entry shape so wallHp has a place to live.
        // Mirrors createTileStateEntry's defaults so nothing downstream
        // reads `undefined` on the previously-empty fields.
        entry = {
          fertility: 0,
          wear: 0,
          growthStage: 0,
          salinized: 0,
          fallowUntil: 0,
          yieldPool: 0,
          nodeFlags: 0,
          lastHarvestTick: -1,
        };
        tileState.set(idx, entry);
      }
      // v0.8.5 Tier 3: gates use gateMaxHp (75) for differentiated HP from walls (50).
      entry.wallHp = newTile === TILE.GATE
        ? Number(BALANCE.gateMaxHp ?? BALANCE.wallMaxHp ?? 50)
        : Number(BALANCE.wallMaxHp ?? 50);
    }
  } else if (oldTile === TILE.WALL || oldTile === TILE.GATE) {
    const tileState = state.grid.tileState;
    if (tileState) {
      const entry = tileState.get(idx);
      if (entry && entry.wallHp != null) {
        // Drop the wallHp field (back to undefined). The rest of the
        // tileState entry stays intact (nodeFlags, yieldPool, etc.) per the
        // setTile preservation pass.
        delete entry.wallHp;
      }
      // v0.8.5 Tier 2 S2: also drop the regen damage timestamp when the
      // wall/gate is removed so a fresh build doesn't inherit stale state.
      if (entry && entry.lastWallDamageTick != null) {
        delete entry.lastWallDamageTick;
      }
    }
  }
}

/**
 * Mutate a tile + run cleanup cascade. Returns true if the tile actually
 * changed type, false if it was already the requested type or out-of-bounds.
 *
 * Prefer this over calling `setTile` directly during simulation. Scenario
 * generation (pre-state-init) can use raw setTile since no agents exist yet.
 *
 * @param {import("../../app/types.js").GameState} state
 * @param {number} ix
 * @param {number} iz
 * @param {number} newTile
 * @returns {boolean}
 */
export function mutateTile(state, ix, iz, newTile) {
  const grid = state?.grid;
  if (!grid) return false;
  const idx = ix + iz * grid.width;
  const oldTile = grid.tiles[idx];
  const changed = setTile(grid, ix, iz, newTile);
  if (!changed) return false;
  onTileMutated(state, ix, iz, oldTile, newTile);
  return true;
}

File: project_utopia/simulation/lifecycle/tile_mutation_hooks.py
"""Tile-mutation cascade cleanup (port of ``TileMutationHooks.js``).

Whenever a tile changes type mid-simulation we must:

1. Refresh the building-count summary (``state["buildings"]``).
2. Release any worker reservation on the tile.
3. Invalidate worker paths / targets that touched the tile (only when the
   *new* tile blocks pathfinding — non-blocking transitions like
   ``GRASS → ROAD`` keep the path valid).
4. Mark the tile key dirty for downstream caches.
5. Optionally seed / clear wall HP when the tile becomes / ceases to be a
   ``WALL`` or ``GATE``.

The Python port also exposes a hook registry so :mod:`world.events` can
wire ``raid → RUINS`` mutations without touching this module directly.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from project_utopia.world.grid import TILE, Grid

__all__ = [
    "BLOCKING_TILES",
    "on_tile_mutated",
    "mutate_tile",
    "register_mutation_hook",
    "clear_mutation_hooks",
    "apply_world_event_impact",
]


BLOCKING_TILES: frozenset[int] = frozenset(
    {TILE["RUINS"], TILE["WALL"], TILE["WATER"]}
)
"""Tiles that block pathfinding. Mutating *to* one of these requires
path invalidation; mutating from one is non-blocking (path stays valid)."""


# Pluggable hook registry. Subscribers receive
# ``(state, ix, iz, old_tile, new_tile)`` after the canonical cascade has
# run. Stored as a list to preserve registration order — see migration
# brief: deterministic = sort by registration order, not callable identity.
_MUTATION_HOOKS: list[Callable[[dict[str, Any], int, int, int, int], None]] = []


def register_mutation_hook(
    hook: Callable[[dict[str, Any], int, int, int, int], None],
) -> Callable[[], None]:
    """Register a post-cascade tile-mutation hook.

    Returns an ``unsubscribe`` callable for symmetry with ``on_event``.
    """
    if hook not in _MUTATION_HOOKS:
        _MUTATION_HOOKS.append(hook)

    def unsubscribe() -> None:
        try:
            _MUTATION_HOOKS.remove(hook)
        except ValueError:
            pass

    return unsubscribe


def clear_mutation_hooks() -> None:
    """Drop all registered hooks. Test-only helper."""
    _MUTATION_HOOKS.clear()


def _rebuild_building_stats(grid: Grid) -> dict[str, int]:
    """Cheap pass over the tile array — same shape as JS ``rebuildBuildingStats``."""
    counts = {
        "warehouses": 0,
        "farms": 0,
        "lumbers": 0,
        "walls": 0,
        "roads": 0,
        "quarries": 0,
        "herbGardens": 0,
        "kitchens": 0,
        "smithies": 0,
        "clinics": 0,
        "bridges": 0,
        "gates": 0,
        "ruins": 0,
    }
    # ``grid.tiles`` is a numpy array (z, x). Flatten with .ravel() so
    # iteration is contiguous and deterministic.
    tiles = grid.tiles.ravel()
    for v in tiles.tolist():
        if v == TILE["WAREHOUSE"]:
            counts["warehouses"] += 1
        elif v == TILE["FARM"]:
            counts["farms"] += 1
        elif v == TILE["LUMBER"]:
            counts["lumbers"] += 1
        elif v == TILE["WALL"]:
            counts["walls"] += 1
        elif v == TILE["ROAD"]:
            counts["roads"] += 1
        elif v == TILE["QUARRY"]:
            counts["quarries"] += 1
        elif v == TILE["HERB_GARDEN"]:
            counts["herbGardens"] += 1
        elif v == TILE["KITCHEN"]:
            counts["kitchens"] += 1
        elif v == TILE["SMITHY"]:
            counts["smithies"] += 1
        elif v == TILE["CLINIC"]:
            counts["clinics"] += 1
        elif v == TILE["BRIDGE"]:
            counts["bridges"] += 1
        elif v == TILE["GATE"]:
            counts["gates"] += 1
        elif v == TILE["RUINS"]:
            counts["ruins"] += 1
    return counts


def _release_tile_reservation(state: dict[str, Any], ix: int, iz: int) -> None:
    reservation = state.get("_jobReservation")
    if reservation is None:
        return
    release_fn = getattr(reservation, "release_tile", None) or getattr(
        reservation, "releaseTile", None
    )
    if callable(release_fn):
        release_fn(ix, iz)


def _invalidate_agent_paths(
    state: dict[str, Any], ix: int, iz: int, new_blocks: bool
) -> None:
    agents = state.get("agents") or []
    for agent in agents:
        if not isinstance(agent, dict):
            continue
        if not bool(agent.get("alive", True)):
            continue
        target = agent.get("targetTile")
        target_matches = (
            isinstance(target, dict)
            and int(target.get("ix", -1)) == ix
            and int(target.get("iz", -1)) == iz
        )
        path_blocked_here = False
        if new_blocks:
            path = agent.get("path")
            if isinstance(path, list):
                for step in path:
                    if (
                        isinstance(step, dict)
                        and int(step.get("ix", -1)) == ix
                        and int(step.get("iz", -1)) == iz
                    ):
                        path_blocked_here = True
                        break
        if not (target_matches or path_blocked_here):
            continue
        if target_matches:
            agent["targetTile"] = None
        agent["path"] = None
        agent["pathIndex"] = 0
        agent["pathGridVersion"] = -1
        desired = agent.get("desiredVel")
        if isinstance(desired, dict):
            desired["x"] = 0.0
            desired["z"] = 0.0
        else:
            agent["desiredVel"] = {"x": 0.0, "z": 0.0}
        bb = agent.get("blackboard")
        if isinstance(bb, dict):
            bb["pendingPathWorkerKey"] = ""
            bb["pendingPathTargetTile"] = None


def on_tile_mutated(
    state: dict[str, Any],
    ix: int,
    iz: int,
    old_tile: int,
    new_tile: int,
) -> None:
    """Run the cascade-cleanup pass after a tile changes type.

    Idempotent — safe to call when ``old_tile == new_tile`` (will early-return).
    """
    if old_tile == new_tile:
        return
    grid = state.get("grid")
    if grid is None:
        return

    # 1. Rebuild building counts so any system that runs later in the same
    # tick sees a fresh layout.
    if isinstance(grid, Grid):
        state["buildings"] = _rebuild_building_stats(grid)

    # 2. Release reservations.
    _release_tile_reservation(state, ix, iz)

    # 3. Invalidate paths / targets when needed.
    new_blocks = int(new_tile) in BLOCKING_TILES
    _invalidate_agent_paths(state, ix, iz, new_blocks)

    # 4. Dirty-key marker for downstream caches (ProcessingSystem, etc.).
    dirty = state.get("_tileMutationDirtyKeys")
    if not isinstance(dirty, set):
        dirty = set()
        state["_tileMutationDirtyKeys"] = dirty
    dirty.add(f"{ix},{iz}")

    # 5. Wall / gate HP lifecycle.
    if isinstance(grid, Grid):
        tile_state = getattr(grid, "tile_state", None)
        if tile_state is not None:
            idx = iz * grid.width + ix
            if new_tile in (TILE["WALL"], TILE["GATE"]):
                entry = tile_state.get(idx)
                if entry is None:
                    entry = {
                        "fertility": 0.0,
                        "wear": 0.0,
                        "growthStage": 0,
                        "salinized": 0.0,
                        "fallowUntil": 0,
                        "yieldPool": 0.0,
                        "nodeFlags": 0,
                        "lastHarvestTick": -1,
                    }
                    tile_state[idx] = entry
                entry["wallHp"] = (
                    75.0 if new_tile == TILE["GATE"] else 50.0
                )
            elif old_tile in (TILE["WALL"], TILE["GATE"]):
                entry = tile_state.get(idx)
                if isinstance(entry, dict):
                    entry.pop("wallHp", None)
                    entry.pop("lastWallDamageTick", None)

    # Fan out to externally-registered hooks (e.g. world.events).
    for hook in list(_MUTATION_HOOKS):
        try:
            hook(state, ix, iz, old_tile, new_tile)
        except Exception:  # pragma: no cover — defensive
            # Hooks are observers, not transactional. A bad hook should
            # never poison the simulation tick.
            continue


def mutate_tile(state: dict[str, Any], ix: int, iz: int, new_tile: int) -> bool:
    """Mutate a tile + run the cleanup cascade. Returns ``True`` on change."""
    grid = state.get("grid")
    if not isinstance(grid, Grid):
        return False
    old_tile = int(grid.get_tile(ix, iz))
    if old_tile == new_tile:
        return False
    grid.set_tile(ix, iz, new_tile)
    on_tile_mutated(state, ix, iz, old_tile, int(new_tile))
    return True


def apply_world_event_impact(
    state: dict[str, Any],
    event_type: str,
    impact_tile: tuple[int, int] | None,
) -> None:
    """Apply tile-mutation side effects from a ``WorldEventSystem`` event.

    This is the hook signature the migration brief asks for: when a raid
    resolves with an ``impactTile`` payload field, mutate the tile to
    ``RUINS`` and fan-out via :func:`on_tile_mutated`.

    Callers (typically :mod:`project_utopia.world.events`) invoke this
    after their per-event lifecycle handler returns.
    """
    if impact_tile is None:
        return
    ix, iz = int(impact_tile[0]), int(impact_tile[1])
    if event_type in ("banditRaid", "bandit_raid"):
        mutate_tile(state, ix, iz, TILE["RUINS"])
    elif event_type == "wildfire":
        # Wildfire burns flammable production tiles down to GRASS.
        mutate_tile(state, ix, iz, TILE["GRASS"])
